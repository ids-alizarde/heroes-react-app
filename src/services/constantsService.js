export const CONSTANTS = {

    login:  '[auth] login',
    logout: '[auth] logout'
};