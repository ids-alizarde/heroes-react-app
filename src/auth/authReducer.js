import { CONSTANTS } from "../services/constantsService";

// const state = {
//     name: 'Aldo',
//     logged: true
// };

export const authReducer = ( state = {}, action ) => {

    switch ( action.type ) {
        case CONSTANTS.login:
            
            return {
                ...action.payload,
                logged: true
            };

        case CONSTANTS.logout:
            
            return {
                logged: false
            };

        default:
            return state;
    }
}