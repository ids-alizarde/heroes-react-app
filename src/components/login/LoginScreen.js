import React, { useContext } from 'react';
import { CONSTANTS } from '../../services/constantsService';
import { AuthContext } from '../../auth/AuthContext';

export const LoginScreen = ({ history }) => {

    const { dispatch } = useContext( AuthContext );
    const lastPath = localStorage.getItem( 'lastPath' ) || '/';

    const handleLogin = () => {
        
        // history.push('/');
        // history.replace('/');

        const action = {
            type: CONSTANTS.login,
            payload: {
                name: 'Aldo Ivan'
            }
        }

        dispatch(action);
        history.replace( lastPath );
    }
        
    return (
        <div className="container mt-5">
            <h1>Login Screen</h1>
            <hr/>

            <button className="btn btn-primary" onClick={ handleLogin }>
                Ingresar
            </button>
        </div>
    )
}
