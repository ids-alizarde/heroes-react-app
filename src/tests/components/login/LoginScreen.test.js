import '@testing-library/jest-dom';
import React from 'react';
import { mount } from 'enzyme';
import { LoginScreen } from '../../../components/login/LoginScreen';
import { AuthContext } from '../../../auth/AuthContext';
import { CONSTANTS } from '../../../services/constantsService';


describe('Pruebas en el componente <LoginScreen />', () => {

    const contextValue = {
        user: { 
            logged: false 
        },
        dispatch: jest.fn()
    }

    const historyMock = {
        replace: jest.fn()
    }

    let wrapper = mount(
        <AuthContext.Provider value={ contextValue }>
            <LoginScreen history={ historyMock }/>
        </AuthContext.Provider>
    );

    beforeEach( () => {
        jest.clearAllMocks();
        wrapper = mount(
        <AuthContext.Provider value={ contextValue }>
            <LoginScreen history={ historyMock }/>
        </AuthContext.Provider>
    );
    });

    test('Debe de mostrar el <LoginScreen /> correctamente', () => {

        expect( wrapper ).toMatchSnapshot();
    });
    
    test('Debe de mandar a llamar el dispatch y realizar la navegacion esperada', () => {

        const action = {
            type: CONSTANTS.login,
            payload: {
                name: 'Aldo Ivan'
            }
        }
        
        wrapper.find( 'button' ).simulate( 'click' );

        expect( contextValue.dispatch ).toHaveBeenCalled();
        expect( contextValue.dispatch ).toHaveBeenCalledWith( action );
        expect( historyMock.replace ).toHaveBeenCalled();
        expect( historyMock.replace ).toHaveBeenCalledWith( '/' );
    });

    test('Debe de mandar a llamar el dispatch y realizar la navegacion esperada y el localStorage', () => {

        const action = {
            type: CONSTANTS.login,
            payload: {
                name: 'Aldo Ivan'
            }
        }

        const handleClick = wrapper.find( 'button' ).prop( 'onClick' );
        handleClick();

        expect( contextValue.dispatch ).toHaveBeenCalledWith( action );

        localStorage.setItem( 'lastPath', '/marvel');
        handleClick();

        expect( historyMock.replace ).toHaveBeenCalledWith( '/marvel' );
        expect( historyMock.replace ).toHaveBeenCalledTimes(2);
    });
})
