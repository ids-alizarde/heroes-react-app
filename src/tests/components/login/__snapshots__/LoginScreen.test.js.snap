// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <LoginScreen /> Debe de mostrar el <LoginScreen /> correctamente 1`] = `
<div
  className="container mt-5"
>
  <h1>
    Login Screen
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Ingresar
  </button>
</div>
`;
